Add CLIMA micro

trontrytel opened the pull request:
This PR adds CLIMA 1-moment micro. See issue #16 

Additional comments
- The Clima 1M microphysics is added as a non-default option. 
- The Clima 1M microphysics uses a temperature-dependent latent heat of fusion and a phase partition dependent latent heat of vaporization L=L(lambda). (This is consistent with TC.jl.)
- For most cases the Seifert-Beheng liquid-only microphysics or the saturation-adjustment only schemes are kept as the defaults. The Seifert-Beheng microphysics implementation uses the temperature-dependent latent heats of fusion and vaporization. 
- This PR also changes the values of specific heats for all microphysics schemes to match those used in CLIMAParameters
- This PR also fixes a minor bug in saturation adjustment routine where the returned phase partitioning was not updated to the last iteration.
- This PR also adds cloud ice to output diagnostics such as cloud cover or fraction.

trontrytel commented:
I'm still debugging TRMM. But in the meantime please take a look at the comparison between the 1-moment clima scheme and the Seifert and Beheng scheme for Dycoms rf02 and Rico. The clima scheme is creating more precipitation for the Sc case but less precipitation for the Cu case. The results seem within the range of differences you would expect from different microphysics schemes (sadly it is a big range). 

Dycoms RF02:
[paths_Dycoms_rf02_les.pdf](https://github.com/CliMA/pycles/files/8101096/paths_Dycoms_rf02_les.pdf)
[profiles_Dycoms_rf02_les.pdf](https://github.com/CliMA/pycles/files/8101097/profiles_Dycoms_rf02_les.pdf)

Rico:
[paths_Rico_les.pdf](https://github.com/CliMA/pycles/files/8101101/paths_Rico_les.pdf)
[profiles_Rico_les.pdf](https://github.com/CliMA/pycles/files/8101102/profiles_Rico_les.pdf)

The question is this though: Do we want to calibrate some of the clima scheme parameters with some of the Seifert and Beheng simulations? Or try to match some of the results described in the two papers better?
Our rain formation function is definitely very simple. I will be testing and improving it more during the summer. It might be good to agree on the datasets you guys need for before summer deadlines. 

cc: @yairchn @ilopezgp @szy21 @charleskawczynski

yairchn commented:
great work. 
How do these compare with the results in the papers?

yairchn commented on Csrc/microphysics_CLIMA.h:
```diff
@@ -257,16 +225,15 @@ void CLIMA_accretion_snow_rain(double q_rai, double q_sno, double rho, double T,
     }
     else{
-      double tmp = pi / rho * CLIMA_n0_sno(q_sno, rho) * n0_rai * m0_rai * Chi_m_rai * E_rai_sno *
+      tmp = pi / rho * CLIMA_n0_sno(q_sno, rho) * n0_rai * m0_rai * Chi_m_rai * E_rai_sno *
```
why did we avoid the `double` assignment ?

trontrytel replied:
I'm defining `tmp` as `double` here: https://github.com/CliMA/pycles/blob/6bb6dab0e87366558f2b52d77721864505e34b49/Csrc/microphysics_CLIMA.h#L170

trontrytel commented:
This PR is ready for review. The TRMM simulations are stable. I'm re-running the Rico and Dycoms RF02 cases right now, to double check. A couple of questions and differences between this PR and CliMA packages:

- PyCLES right now has two formulations for latent heat: (i) some empirical fit that depends on the temperature, (ii) a phase partition function weighted sum of Lv and Ls (Lv and Ls are constant in that case). This is different from CliMA packages where we use the phase partition function but also take Lv and Ls to depend on the temperature.

- I have implemented the latent heat of fusion as `LH_f0 + (cl - ci) * (T - Tt)` which is consistent with CliMA packages, but maybe not with the treatment of latent heats in PyCLES. 

- The values of specific heats for liquid and ice (`cl`, `ci`) are slightly different between PyCLES and CliMA packages. 

- I'm not sure about the formulation of the entropy source term from melting snow. I followed what is done in the arctic microphysics. The [2015 paper](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1002/2015MS000496) formulates it differently by using the freezing level.

trontrytel commented:
I also ran a few test simulations to check how the results depend on some of the parameters. The plots are below. 
- The first page shows the behaviour with the settings from the current PR. 
- Second page looks at how the results change with the rain autoconversion timescale (tau), temperature below which there is only ice (T cold) and latent heat formulation. 
- The third page looks at snow autoconversion timescale and a different snow autoconversion formulation that depends on supersaturation.  

[TRMM_sensitivity.pdf](https://github.com/CliMA/pycles/files/8180757/TRMM_sensitivity.pdf)

This is very pre-eliminary quick look into the simulations. But the bottom line is that there is a lot of variability. If we want to use the scheme to generate meaningful data, it would be great to first calibrate the most uncertain parameters with some targeted cases.

ilopezgp commented:
> This PR is ready for review. The TRMM simulations are stable. I'm re-running the Rico and Dycoms RF02 cases right now, to double check. A couple of questions and differences between this PR and CliMA packages:
> 
> * PyCLES right now has two formulations for latent heat: (i) some empirical fit that depends on the temperature, (ii) a phase partition function weighted sum of Lv and Ls (Lv and Ls are constant in that case). This is different from CliMA packages where we use the phase partition function but also take Lv and Ls to depend on the temperature.
> * I have implemented the latent heat of fusion as `LH_f0 + (cl - ci) * (T - Tt)` which is consistent with CliMA packages, but maybe not with the treatment of latent heats in PyCLES.
> * The values of specific heats for liquid and ice (`cl`, `ci`) are slightly different between PyCLES and CliMA packages.
> * I'm not sure about the formulation of the entropy source term from melting snow. I followed what is done in the arctic microphysics. The [2015 paper](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1002/2015MS000496) formulates it differently by using the freezing level.

- About the latent heat formulation: I would go with the latter (a phase partition function weighted sum of Lv and Ls (Lv and Ls are constant in that case)). This may not be consistent with what we do in ClimaMicrophysics, but it is what we do in TC.jl: for example, in the [buoyancy gradient closure](https://github.com/CliMA/TurbulenceConvection.jl/blob/main/src/closures/buoyancy_gradients.jl#L30).

- About the latent heat of fusion - I would say whatever is consistent with TC.jl and what we want to do in CliMA.
- For the specific heats for liquid and ice, I would also suggest following the Clima parameters.
- Not sure about the entropy source rationale (@tapios)?

ilopezgp commented:
> I also ran a few test simulations to check how the results depend on some of the parameters. The plots are below.
> 
> * The first page shows the behaviour with the settings from the current PR.
> * Second page looks at how the results change with the rain autoconversion timescale (tau), temperature below which there is only ice (T cold) and latent heat formulation.
> * The third page looks at snow autoconversion timescale and a different snow autoconversion formulation that depends on supersaturation.
> 
> [TRMM_sensitivity.pdf](https://github.com/CliMA/pycles/files/8180757/TRMM_sensitivity.pdf)
> 
> This is very pre-eliminary quick look into the simulations. But the bottom line is that there is a lot of variability. If we want to use the scheme to generate meaningful data, it would be great to first calibrate the most uncertain parameters with some targeted cases.

Did you fix the random number in these experiments? It would be good if you can compare the spread from changing parameters to the spread from stochasticity. In my experience, TRMM can vary quite a lot if you just use a different random number.

trontrytel commented:
> Did you fix the random number in these experiments? It would be good if you can compare the spread from changing parameters to the spread from stochasticity. In my experience, TRMM can vary quite a lot if you just use a different random number.

Yes, the random seed is fixed. I know that TRMM has a lot of spread there. I can run some ensemble simulations, but I'm expecting a similar amount of spread as seen in the arctic lines. 

If we decide on some more limited set of parameters we want to test based on the plots I have shown so far, I can run a sensitivity experiment with ensembles. But maybe a good thing to think about first would be to choose the cases and parameters we want to calibrate. TRMM might not be the best case to test and calibrate to

szy21 commented on Csrc/microphysics_CLIMA.h:
```diff
@@ -172,65 +466,111 @@ void CLIMA_microphysics_sources(const struct DimStruct *dims, struct LookupStruc
-                    rate = fmax(1.05 * qr_tendency_tmp * dt_ / (-fmax(qr_tmp, microph_eps)), rate);
+                    // ... check the source term magnitudes ...
+                    ql_tendency_tmp = -qr_tendency_aut + ql_tendency_acc;
+                    qi_tendency_tmp = -qs_tendency_aut + qs_tendency_acc;
```
```suggestion
                    qi_tendency_tmp = -qs_tendency_aut + qi_tendency_acc;
```

szy21 commented on Csrc/microphysics_CLIMA.h:
```diff
@@ -278,31 +616,37 @@ void CLIMA_entropy_source_formation(const struct DimStruct *dims, struct LookupS
 
-                entropy_tendency[ijk] += S_P         * 0.5 * (qr_tendency[ijk] + fabs(qr_tendency[ijk])) -
-                                         (S_E + S_D) * 0.5 * (qr_tendency[ijk] - fabs(qr_tendency[ijk]));
+                const double tmp = S_P * precip_formation_rate[ijk] - (S_E + S_D) * evaporation_sublimation_rate[ijk];
```
What is this line for? (Maybe you want `entropy_tendency[ijk] += tmp` in the following line?)

trontrytel replied:
Thats a leftover from when I was printing NaNs left and right 😂 We don' t need the tmp. The entropy tendency is updated anyway

szy21 commented on Csrc/thermodynamics_sa.h:
```diff
@@ -61,30 +61,31 @@ void eos_c(struct LookupStruct *LT, double (*lam_fp)(double), double (*L_fp)(dou
         *qv = qv_star_2;
-        *ql = lam_2 * sigma_2;
-        *qi = (1.0 - lam_2) * sigma_2;
+        *ql = lam_fp(T_2) * sigma_2;
+        *qi = (1.0 - lam_fp(T_2)) * sigma_2;
```
Do we need to change this two lines? I think `lam_2` is the same as `lam_fp(T_2)`, but I'm fine either way.

trontrytel replied:
I think it needs to be updated here

szy21 reviewed (approved):
I tried to go over it but didn't read every single line:) I think it looks good to me.

trontrytel commented:
I have changed the values of specific heats to match the ones we use in CliMA and did not change anything about the latent heat formulation. I think this is the most consistent path to take while working with both models. 

I have put all the additional microphysics parameters into `generate_parameters.py` file. I have kept the `generate_namelist.py` file as is. In the 3 cases I have looked at so far (Dycoms_RF02, Rico and TRMM_LBA) I have left commented out the changes needed to run the model with the clima 1-moment scheme. This way the new code should not change anyones default simulations. 

The final comparison between the Seifert-Beheng scheme for Dycoms_RF02 and Rico and arctic scheme for TRMM_LBA shows differences. This is of course to be expected. Before we start using the new scheme more broadly, we should test in on other cases and consider some calibration of the parameters. This can be done in next PRs.

Here are the comparison plots:
[LES_paths_Dycoms_rf02.pdf](https://github.com/CliMA/pycles/files/8248131/LES_paths_Dycoms_rf02.pdf)
[LES_profiles_Dycoms_rf02.pdf](https://github.com/CliMA/pycles/files/8248132/LES_profiles_Dycoms_rf02.pdf)

[LES_paths_Rico.pdf](https://github.com/CliMA/pycles/files/8248134/LES_paths_Rico.pdf)
[LES_profiles_Rico.pdf](https://github.com/CliMA/pycles/files/8248135/LES_profiles_Rico.pdf)

[LES_paths_TRMM.pdf](https://github.com/CliMA/pycles/files/8248137/LES_paths_TRMM.pdf)
[LES_profiles_TRMM.pdf](https://github.com/CliMA/pycles/files/8248138/LES_profiles_TRMM.pdf)

trontrytel commented:
bors try

ilopezgp commented on Csrc/microphysics_CLIMA.h:
```diff
@@ -4,155 +4,376 @@
+  // accretion qr ql
+  if(q_rai > microph_eps && q_liq > microph_eps){
+    tmp =
+        q_liq * E_liq_rai * n0_rai * a0_rai * CLIMA_v0_rai(rho) * Chi_a_rai * Chi_v_rai / lambda_rai *
+        tgamma(ae_rai + ve_rai + Delta_a_rai + Delta_v_rai + 1.) /
+        pow(lambda_rai * r0_rai, ae_rai + ve_rai + Delta_a_rai + Delta_v_rai);
```
In a later PR perhaps, it would be good to create a function for the abstract function for any condensate and hydrometeor, like equation (A7) in your paper. It will make code checking easier, since there is a lot of functional form repetition here.

trontrytel replied:
This is how it is coded in the Julia implementation. I don't want to do a PyCLES code refactor now, so I followed whatever coding practices were there. There is a lot of repetition in the implementations of all the microphysics schemes.

ilopezgp commented on Csrc/microphysics_CLIMA.h:
```diff
@@ -4,155 +4,376 @@
+    double lambda_sno = CLIMA_lambda(q_sno, rho, CLIMA_n0_sno(q_sno, rho), m0_sno, me_sno, r0_sno, Chi_m_sno, Delta_m_sno);
+
+    *qs_tendency_dep_sub =
+      4. * pi * CLIMA_n0_sno(q_sno, rho) / rho * S * G / pow(lambda_sno, 2) *
+      (a_vent_sno + b_vent_sno * pow(nu_air / D_vapor, 1./3.) /
+        pow(r0_sno * lambda_sno, (ve_sno + Delta_v_sno) / 2.) *
+        pow(2. * v0_sno * Chi_v_sno / nu_air / lambda_sno, 1./2.) *
+        tgamma((ve_sno + Delta_v_sno + 5.) / 2.)
+      );
```
Is this equation (A11) in your draft? If so, are you defining S (code) = S-1 (paper)?

trontrytel replied:
Yes: https://github.com/CliMA/pycles/blob/c06be92c844a2db29c4a279b9e203bcdc8123ace/Csrc/microphysics_CLIMA.h#L251

ilopezgp commented on Csrc/microphysics_CLIMA.h:
```diff
@@ -4,155 +4,376 @@
 
-                qr_velocity[ijk] = interp_2(qr_velocity[ijk], qr_velocity[ijk+1]) ;
+    *qs_tendency_melt =
+      -4. * pi * CLIMA_n0_sno(q_sno, rho) / rho * K_therm / L_f * (T - Tf) / pow(lambda_sno, 2) *
+      (a_vent_sno + b_vent_sno * pow(nu_air / D_vapor, 1./3.) /
+        pow(r0_sno * lambda_sno, (ve_sno + Delta_v_sno) / 2.) *
+        pow(2. * v0_sno * Chi_v_sno / nu_air / lambda_sno, 1./2.) *
+        tgamma((ve_sno + Delta_v_sno + 5.) / 2.)
```
Should (A12) in your draft include the negative sign, as here? Both are defining the sink to snow due to melting, but have opposite signs.

trontrytel replied:
I guess you are right. (A12) returns a positive number and should be subtracted when applied to snow tendencies. 

I always have a nomenclature/sign problem when defining sources and sinks that for example are always positive or always negative.

ilopezgp commented on Csrc/microphysics_CLIMA.h:
```diff
@@ -4,155 +4,376 @@
+              double qs_tmp = fmax(qs[ijk], 0.0);
+
+              qr_velocity[ijk] = -fmin(CLIMA_terminal_velocity_rai(density[k], qr_tmp), 10.0);
+              qs_velocity[ijk] = -fmin(CLIMA_terminal_velocity_sno(density[k], qs_tmp), 10.0);
```
Is the minimum terminal velocity of 10 m/s in the draft? Should we expose it as a parameter?

trontrytel replied:
This is not how it is implemented in the Julia version of the scheme. I followed the limitation practices that I found in PyCLES.

ilopezgp commented:
My understanding of this PR is:

- Adds the Clima 1M microphysics as a non-default option. The Clima 1M microphysics uses a temperature-dependent latent heat of fusion and a phase partition dependent latent heat of vaporization `L=L(lambda)`. This is consistent with TC.jl.
- Corrects the reference values of latent heat for all microphysics schemes.
- Maintains the Seifert-Beheng liquid-only microphysics scheme as the default. The Seifert-Beheng microphysics scheme uses temperature-dependent latent heats of fusion and vaporization. This does not match TC.jl, but matches the approach taken to generate the Stratocumulus-to-Cumulus LES data in Shen et al (2022).

Let's update the PR description with this and anything else that you did in this PR, @trontrytel . I suppose the next step is to test further the CliMA 1M micro before we switch to that for the deep convective cfSite runs? Thinking about the end goal here, can we make the claim that the Stratocumulus-to-Cumulus LES data in Shen et al (2022) produced with SB microphysics would not change much if we changed to CliMA 1M? Do you envision any inconsistencies if we try to learn from SB-based Stratocumulus-to-Cumulus LES and CliMA 1M-based deep convection? @trontrytel @szy21 @tapios

trontrytel commented:
>I suppose the next step is to test further the CliMA 1M micro before we switch to that for the deep convective cfSite runs? 

Yes. That's why I did not change it to be the default.

>Thinking about the end goal here, can we make the claim that the Stratocumulus-to-Cumulus LES data in Shen et al (2022) >produced with SB microphysics would not change much if we changed to CliMA 1M? 

I think we were using the 1-moment scheme there. In general I think we should expect differences every time we switch a microphysics scheme. 

> Do you envision any inconsistencies if >we try to learn from SB-based Stratocumulus-to-Cumulus LES and CliMA 1M based deep convection? @trontrytel @szy21 >@tapios

It is a good question. I think the SB scheme is better and if possible we should use it. For calibrating TC.jl it could be worth it to use both. Calibrating to a smarter scheme (SB) would give us a better performing parameterisation. Calibrating to data generated with the same scheme in LES and TC.jl would give us more understanding of the biases we still have in the model and parameterisation coupling.

szy21 commented:
Actually, all the Stratocumulus-to-Cumulus LES data are produced with liquid only CliMA 1M (the one that Anna coded before). I did try some SB microphysics at the very beginning and the results are sensitive to the microphysics scheme, especially at the cloud top, if I remember correctly.

trontrytel commented:
I'll update the PR description once I fix the parameter naming conflict discovered by the CI. 

Aside from the list from @ilopezgp I have:

- fixed a minor bug in phase partitioning returned by saturation adjustment (the phase partitioning was not based on the last scheme iteration)
- added some missing diagnostics like including cloud ice in cloud cover diagnostics and outputting the ice water path, etc.

ilopezgp commented:
> Actually, all the Stratocumulus-to-Cumulus LES data are produced with liquid only CliMA 1M (the one that Anna coded before). I did try some SB microphysics at the very beginning and the results are sensitive to the microphysics scheme, especially at the cloud top, if I remember correctly.

Ok, this is much better then! All problems solved! Then, we just have to continue testing the CliMA 1M for deep convection, and once we make the change everything will be consistent. Woohooo!!

szy21 commented:
> I'll update the PR description once I fix the parameter naming conflict discovered by the CI.

Let me know if you can't fix it. I'll take a look locally. It could be there is some difference in the package versions.

trontrytel commented:
> > I'll update the PR description once I fix the parameter naming conflict discovered by the CI.
> 
> Let me know if you can't fix it. I'll take a look locally. It could be there is some difference in the package versions.

I found it. Apparently there is a conflict in names with one of the arctic parameters. I don't understand why it does not show when I compile locally. But I will just have all my new parameters start with `CLIMA_`. This way there will be no conflicts for sure.

trontrytel commented:
bors try

trontrytel commented:
bors r+
